surveys/micro/2025-06-compiler-performance.md

```diff
@@ -0,0 +1,224 @@
+- `cargo test`
+
+Frequency:
+
+- Rarely
+- Sometimes
+- Often
+
+### How long do you need to wait for the compiler to rebuild your code after making a change?
+
+Please select the longest duration range amongst the projects that you regularly work on.
+
+Type: select one (optional)
+
+- Less than a second
+- Between 1 and 5 seconds
+- Between 5 and 10 seconds
+- Between 10 and 30 seconds
+- Between 30 seconds and 1 minute
+- Between 1 and 5 minutes
+- More than 5 minutes
+
+### Which of the following problems do you most struggle with?
+
+Type: matrix (optional)
+
+Commands:
+
+- Waiting for a rebuild after making a small change.
+- Waiting for CI workflows that build Rust code.
+- `cargo check` and `cargo build` not sharing compilation cache.
+- `cargo check` and `cargo clippy` not sharing compilation cache.
+- Waiting for an IDE to show me inline error/warning annotations.
+
+Priority:
+
+- Not a problem for me
+- Could be improved
+- Big blocker for me
+
+## Workarounds
+
+### Do you use any of the following compiler options to improve compilation performance?
+
+Type: select all that apply (optional)
+
+- Disable (or reduce) debuginfo (e.g. set `debug = 0` in `Cargo.toml`)
+- Parallel compiler frontend (pass `-Zthreads=<N>` to the compiler)
+- Cranelift codegen backend (e.g. set `codegen-backend = "cranelift"` in `Cargo.toml`)
+- Alternative linker (e.g. `lld`/`mold`/`wild`)
+- Caching compiler wrapper (e.g. `sccache`)
+- Something else (open response)
+
+### If you use an alternative linker, which one do you use?
+
+Type: select all that apply (optional)
+
+- gold
+- lld
+- mold
+- wild
+- Other (open response)
+
+### Do you use any of the following compiler options to improve compilation performance?
+
+Type: select all that apply (optional)
+
+- Disable (or reduce) debuginfo (e.g. setting `debug = 0` in `Cargo.toml`)
+- Parallel compiler frontend (passing `-Zthreads=<N>` to the compiler)
+- Cranelift codegen backend (e.g. setting `codegen-backend = "cranelift"` in `Cargo.toml`)
+- Alternative linker (e.g. LLD/mold/wild)
+
+### Do you use a global `config.toml` override?
+
+You can create a `config.toml` file in your `CARGO_HOME` directory (e.g. `~/.cargo/config.toml`) which can be used to apply certain compilation settings (e.g. using a faster linker) across all Cargo projects on your computer.
+
+Type: select one (optional)
+
+- Yes
+- No
+
+### Do you use any of the following methods to improve compilation performance?
+
+Please select only methods that you have used primarily in order to improve compilation performance, not for other reasons.
+
+Type: select all that apply (optional)
+
+- Splitting crates into smaller crates
+- Reducing the amount of dependencies
+- Disabling default Cargo features of dependencies
+- Creating a Cargo feature that makes certain dependencies (or their features) optional
+- Something else (open response)
+
+### Do you use a nightly compiler to achieve better compilation performance?
```

Comment:
Is nightly faster? (This question surprised me, I'm not sure what to make of it.)

Reply:
> Is nightly faster? (This question surprised me, I'm not sure what to make of it.)

So, one thing is that many people *believe* that it is faster, so it's good to see how large this perception is. The second thing is that while historically this was mostly untrue, recently `nightly` was indeed faster by default, for example for the past year it used the LLD linker by default. So just changing stable to nightly provided speedups.

But the larger effect is that nightly enables you to use more flags and compiler options to actually make compilation faster, even if it is not the default.